### DISCUSSION TITLE
route matching issue

### DISCUSSION THREAD

**Discussion by CovertError** [livewire/livewire#5526] [7. Bugs (archived)]
#5490 this issue still exits when going to other locale's like /ar but for some reason /en works

**Comment by joshhanley** [comment]
@CovertError can you just confirm what your set up is? As the route match PR fixed multiple issues. If you could create a small example repo which reproduces the issue with as little code as possible so we can investigate it too, that would be great! Thanks!

**Reply by CovertError** [reply]
I'm not sure how I didn't see this I'n sorry for the very late reply I'm using mcamara/laravel-localization if you still want i can create an example project

**Reply by joshhanley** [reply]
@CovertError no worries! All good, thanks! I think I've got it covered with the demo repo provided by @voicecode-bv. I've also been able to replicate the issue with `mcamara/laravel-localization`.

**Comment by voicecode-bv** [comment]
We've got the same issue on our end. Had to revert back to v2.10.8 in order to get everything working again.

**Reply by joshhanley** [reply]
@voicecode-bv can you just confirm what your setup is and what version of Livewire you were on? If you could share your code or create a small example repo which reproduces the issue with as little code as possible so we can investigate it too, that would be great! Thanks!

**Reply by voicecode-bv** [reply]
Hi @joshhanley, I'm working on it. It's a very strange issue as it is actually really hard to reproduce. When I have it, I'll let you know.

Working with the latest version of Laravel, latest version of Livewire, on every post request it throws a 404 when for example there's a /nl/ prefix in the url. When the prefix is just / there's no issue. 

/nl/favorieten doesn't work
/favorites works 

After downgrading to Livewire 2.10.8 it's working again

**Reply by voicecode-bv** [reply]
@joshhanley I guess I'm on to something. I've shared a [public repo](https://github.com/voicecode-bv/livewire-test) for you to take a look at. I think it has something to do with the translate helper in web.php. Please see the readme file for installation instructions.

**Reply by joshhanley** [reply]
@voicecode-bv great, thanks! I will try and have a look in the next couple of days 🙂

**Reply by joshhanley** [reply]
@voicecode-bv I've just had a quick look and it seems using `/nl` prefix is fine, I believe the issue is to do with the dynamic route path using the translation helper.

If I change `Route::get(__('routes.favorites'), ...` to `Route::get('favourites', ...` and browse to `/nl/favorites` it all works happily.

 I will have to do some digging to find out why that's happening.

**Reply by voicecode-bv** [reply]
Thanks @joshhanley, appreciate it!

**Comment by joshhanley** [comment]
@voicecode-bv ok, so it is as I suspected, the issue is the route path translation.

Livewire needs to be able to look up the original route that the component was initially loaded on, to ensure the correct middleware are applied to any of Livewire's subsequent requests. This feature is known as peristent middleware, see here in the docs https://laravel-livewire.com/docs/2.x/security#persistent-middleware

The problem is occurring because when Livewire triggers a subsequent request to the `/livewire/messages` endpoint, there is no locale prefix on the path. And because Livewire has accessed a route (the `/livewire/messages`), all route translations are in the default language, so in your example repo they're in English.

This is because whatever localisation package you're using (or your custom implementation in the example repo) can't find any locale prefix to use, so when `LocalizationHelper::setLocale()` is called in the `routes/web.php` file, it assumes the default language.

Then when Livewire tries to match the original route, say `/nl/favorieten`, to see if it can collect the middleware from it, it can't find any routes that match that path, as the router is currently storing it as `/en/favorites`. Hence the 404 error.

So I've dug further into this to see if there is a simple way to fix this. The locale from the original route is actually sent along with the Livewire request. So I thought, if we could some how manually set the locale based of this, then it should all work.

But as I mentioned before, the router already has the routes stored in memory translated into the default locale. So even manually setting the app locale won't help.

As it stands, I'm not sure how to solve this issue. I had ideas of somehow rebooting the router with the correct routes, but that then includes running through all service providers again to ensure all routes are recollected, as for example, the `RouteServiceProvider` is where the `web.php` file is included, but this call could be in any provider such as the `AppServiceProvider` or even service provides from packages.

I've made a note of this to ensure we have support for this in V3, as we can fundamentally change how it works. But I don't know if this is something we're going to be able to fix in V2.

You may ask, why did this suddenly stop working after v2.10.8? That is because the persistent middleware code was changed which fixed a security bug, as such we can't just revert the fix, as we would normally do.

We're open to ideas though if anyone has more experience with how the translation packages work and can find a way to get us a list of routes with the desired translation applied for us to match against.

**Reply by seche** [reply]
Yeah this has become an issue for us due to translatable routes. Had to lock livewire to version 2.10.7 for things to work again. Couldn't find an easy fix. Might need to look into persistent middleware that can apply the locale prior to the livewire route calls...

**Reply by joshhanley** [reply]
@seche yep, it's a pain. Are you using any localisation package?

I found the `mcamara/laravel-localization` loads the locale using either a prefix or the browser's or using the `Accept-Language` header from the browser. So I couldn't get Livewire to just set the app's language based on the previous request, which the package could then use. That would have been a nice generic solution to prevent this, but it doesn't work as it's too package specific.

I will see if I can get a demo middleware working that people can add to their apps and adapt based on the localisation package they are using.

@voicecode-bv and @CovertError are either of you using a localisation package and if so, what?

**Reply by voicecode-bv** [reply]
@joshhanley no, not using a package here as all the logic is within the single helper class I used.

**Reply by joshhanley** [reply]
@voicecode-bv ah ok, thanks. Wasn't sure if that was your code or just an extraction from the localisation package (as it's very similar) 🙂

I'll see if I can come up with something that works for both.

**Reply by seche** [reply]
@joshhanley yes, we use the `mcamara/laravel-localization` translated routes. So when you are on the English routes and the locale is not set it defaults to English so it still _works_ but if you're using any other locale you get the error because locale is not set on the request.

I don't know but will dig a bit deeper if I can find some time.

**Reply by joshhanley** [reply]
@seche ok thanks for confirming!

Don't dig just yet, I think I may have a fix for it. Just doing some further testing now.

**Reply by voicecode-bv** [reply]
Thanks for your efforts @joshhanley, really appreciate it!

**Reply by joshhanley** [reply]
@voicecode-bv no worries! 🙂

**Comment by joshhanley** [comment]
@CovertError @voicecode-bv and @seche I've submitted a PR #5577 which hopefully will fix this issue. Can someone pull this change and let me know if it fixes the issue? Thanks!

**Comment by joshhanley** [comment]
Thanks to @seche's feedback, I've worked out that the fix in PR #5577 doesn't work when combine with `mcamara/laravel-localization`.

So I don't think that PR is going to work the way it is at the moment.

@CovertError @voicecode-bv and @seche I have another solution, which I think is actually simpler and more robust, but requires modifications in your apps, so I'd like some feedback if you can.

What I'm proposing is to add a method to Livewire's core called `getPayloadLocale()` or something like that (open to naming ideas).

How this would be used in your apps, is wherever you call your `setLocale()` method (whether it be from a locale package method or custom method), you'd pass in the result of the above method.

So if you're using `mcamara/laravel-localization`, you'd do the following:

```php
Route::prefix(LaravelLocalization::setLocale(Livewire\Livewire::getPayloadLocale()))
    ->group(function () {
        Route::get(__('routes.welcome'), Welcome::class);
    });
```

And @voicecode-bv you would do:

```php
Route::prefix(LocalizationHelper::setLocale(Livewire\Livewire::getPayloadLocale()))
    ->group(function () {
        Route::get(__('routes.welcome'), Welcome::class);
    });
```

What it does is, if it is a Livewire request, it will return the locale from the component payload from the initial request.

If it's not a Livewire request, it will just return `null`, which is the default parameter that `setLocale()` accepts anyway, so it would just follow its normal process.

If you want to test it out, run the following:

```bash
composer config repositories.livewire/livewire vcs https://github.com/joshhanley/livewire

composer require livewire/livewire:dev-add-payload-locale-method
```

Then modify your `routes/web.php` to include `Livewire\Livewire::getPayloadLocale()` as per the above.

Let me know what you think of that and whether the solution works.

**Reply by seche** [reply]
Confirming that this works for the `mcamara/laravel-localization`, the updated `getPayloadLocale()` and putting the locale value in the `setLocale()` does the trick. Will be tricky to explain in the docs but will hopefully save someone else some time.

Thanks @joshhanley, great work!

**Comment by joshhanley** [comment]
@CovertError @voicecode-bv and @seche sorry to keep tagging you all!

I believe I have a permanent fix for this translated route paths issue. See PR #5589 for details.

We've opted to go with a designated locale prefix for now (a configurable prefix would be a better option, so we are exploring that possibility for V3).

This won't require any changes in your apps and should just work.

If you want to test it out, run the following:

```bash
composer config repositories.livewire/livewire vcs https://github.com/joshhanley/livewire

composer require livewire/livewire:dev-add-locale-prefix
```

Which will install my `add-locale-prefix` branch. You shouldn't need to make any other changes.

I'm going to merge and tag this PR, probably tomorrow, once I have approval from Caleb.

Hope this helps!

**Reply by voicecode-bv** [reply]
@joshhanley I just tested your dev-add-locale-prefix branch and I think you nailed it. 🔥 Thanks for all your efforts!

**Reply by joshhanley** [reply]
@CovertError great! Thanks for confirming. No worries! 🙂

**Comment by joshhanley** [comment]
Fix for this has been merged and released in v2.12.2.
